util/dao/src/lib.rs

```diff
@@ -1,256 +1,660 @@
+        DaoCalculator {
+            consensus,
+            store,
+            data_loader,
+        }
+    }
+
+    pub fn primary_block_reward(&self, target: &Header) -> Result<Capacity, FailureError> {
+        let target_epoch = self
+            .store
+            .get_block_epoch_index(target.hash())
+            .and_then(|index| self.store.get_epoch_ext(&index))
+            .ok_or(Error::InvalidHeader)?;
+
+        target_epoch.block_reward(target.number())
+    }
+
+    pub fn secondary_block_reward(&self, target: &Header) -> Result<Capacity, FailureError> {
+        if target.number() == 0 {
+            return Ok(Capacity::zero());
+        }
+        let target_parent_hash = target.parent_hash();
+        let target_parent = self
+            .store
+            .get_block_header(target_parent_hash)
+            .ok_or(Error::InvalidHeader)?;
+        let target_epoch = self
+            .store
+            .get_block_epoch_index(target.hash())
+            .and_then(|index| self.store.get_epoch_ext(&index))
+            .ok_or(Error::InvalidHeader)?;
+
+        let target_g2 = calculate_g2(
+            target.number(),
+            &target_epoch,
+            self.consensus.secondary_epoch_reward(),
+        )?;
+        let (_, _, target_parent_u) = extract_dao_data(target_parent.dao())?;
+        let (_, target_c, _) = extract_dao_data(target.dao())?;
+        let reward = u128::from(target_g2.as_u64()) * u128::from(target_parent_u.as_u64())
+            / (max(u128::from(target_c.as_u64()), 1));
+        Ok(Capacity::shannons(reward as u64))
+    }
+
+    pub fn dao_field(
+        &self,
+        rtxs: &[ResolvedTransaction],
+        parent: &Header,
+    ) -> Result<Bytes, FailureError> {
+        let parent_epoch = self
+            .store
+            .get_block_epoch_index(parent.hash())
+            .and_then(|index| self.store.get_epoch_ext(&index))
+            .ok_or(Error::InvalidHeader)?;
+
+        // Freed occupied capacities from consumed inputs
+        let freed_occupied_capacities =
+            rtxs.iter().try_fold(Capacity::zero(), |capacities, rtx| {
+                self.input_occupied_capacities(rtx)
+                    .and_then(|c| capacities.safe_add(c).map_err(Into::into))
+            })?;
+
+        // Newly added occupied capacities from outputs
+        let added_occupied_capacities =
+            rtxs.iter().try_fold(Capacity::zero(), |capacities, rtx| {
+                rtx.transaction
+                    .outputs()
+                    .iter()
+                    .try_fold(Capacity::zero(), |tx_capacities, output| {
+                        output
+                            .occupied_capacity()
+                            .and_then(|c| tx_capacities.safe_add(c))
+                    })
+                    .and_then(|c| capacities.safe_add(c))
+            })?;
+
+        let (parent_ar, parent_c, parent_u) = extract_dao_data(parent.dao())?;
+
+        let parent_g2 = calculate_g2(
```

Comment:
This block of code can be simplified by reusing the logic to calculate base reward:

- First get the reward target of the parent block.
- If target is Some, get its base reward, otherwise `unwrap_or_default`.

```
let g = consensus.finalize_target(parent.number()).map(|target|
  self.base_block_reward(&target)
).unwrap_or_default();
```

The method `base_block_reward` does not exist yet, but is a method I recommend to add. It just returns the sum of primary and secondary reward. I have searched the code base, we never only need one of them.

Reply:
I'm not sure if we can simplify the code here: while we have `g = g1 + g2`, and `g1` is the primary block reward, `g2` is not the secondary block reward. It's only that part of `g2`, depending on the  utilization then, can be counted as secondary block reward.

So while we might argue if we should rename `secondary_block_reward` method to `secondary_miner_block_reward`, I don't think we can simplify the code here to simply use `base_block_reward`.

Reply:
Good catch